### ISSUE TITLE
TLS Handshake Fails on Marshmallow

### ISSUE THREAD

**Issue by WoefulDerelict** [iiordanov/remote-desktop-clients#57]
When attempting to connect to a VNC server via bVNC Pro on my Nexus 9 running Marshmallow I receive the following error: Connection to VNC server failed with reason: [Ljava.lang.StackTraceElement;@1b450f2TLS handshake failed java.net.ssl.SSLHandshakeException: Handshake failed

I had no issues before the upgrade to Marshmallow.

**Comment by WoefulDerelict** [comment]
The host machine contains the following error in response to the failed connection: TLS Handshake failed: The TLS connection was non-properly terminated.

**Comment by iiordanov** [comment]
Could this be an issue with your SSL certificate or perhaps with the
time/date on your device and/or server?

On Fri, Oct 9, 2015 at 12:14 PM, Llewelyn Trahaearn <
notifications@github.com> wrote:

> The host machine contains the following error in response to the failed
> connection: TLS Handshake failed: The TLS connection was non-properly
> terminated.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/iiordanov/remote-desktop-clients/issues/57#issuecomment-146914216>
> .
>



-- 
The conscious mind has only one thread of execution.

**Comment by WoefulDerelict** [comment]
As both devices syncronize their clocks with network sources and other time based services involving both devices like Google Authenticator continue to work without issue I do not believe it is related to either device's clock being innacurate.  I will have to check the SSL certificate more closely to eliminate it as the cause but, I have no issues with other clients so it also seems unlikely.

**Comment by WoefulDerelict** [comment]
The behaviour appears to be consistent for all the VNC servers I was once able to connect to on my Nexus 9 under Lollipop. All of them remain responsive to other clients so this appears to be isolated to bVNC on Marshmallow.

**Comment by mzhou** [comment]
I dug into it and the error that comes from the SSL library is the same as this (at least connecting to vino-server): https://code.google.com/p/android-developer-preview/issues/detail?id=3063

I worked around it by hard coding not supporting auth type 18 and just doing plain auth (which would be nice to have as an option even if this weren't a bug) since I connect through an SSH tunnel anyway.

**Comment by iiordanov** [comment]
Hey,The bug report you posted ends quite inconclusively. Since i don't need a Marshmallow device, i can't hope to fix this if there is no ready path to a solution. If you can help with more suggestions and are willing to test, we can try.iordan                                                                                                                                                                                                                                                                                                                                        --The conscious mind has only one thread of execution.                                                                                                                                                                                                                From: mzhouSent: Monday, October 12, 2015 10:31To: iiordanov/remote-desktop-clientsReply To: iiordanov/remote-desktop-clientsCc: iiordanovSubject: Re: [remote-desktop-clients] TLS Handshake Fails on Marshmallow (#57)I dug into it and the error that comes from the SSL library is the same as this (at least connecting to vino-server): https://code.google.com/p/android-developer-preview/issues/detail?id=3063

I worked around it by hard coding not supporting auth type 18 and just doing plain auth (which would be nice to have as an option even if this weren't a bug) since I connect through an SSH tunnel anyway.

—Reply to this email directly or view it on GitHub.

**Comment by iiordanov** [comment]
Bad news, everyone. Looks like Anonymous Diffie Hellman (DH) ciphers were dropped past API 22 (Android 6.0 is API 23). This means that any protocol like the AnonTLS protocol used by Vino and x11vnc in certain modes of operation is UNSUPPORTED in Android 6 and newer.

http://developer.android.com/reference/javax/net/ssl/SSLEngine.html

There is a workaround for x11vnc if you're running it yourself on the command-line. You need to add the "-vencrypt nodh:only -ssl" options. For example, to automatically find/create a session for your user and use VeNCrypt with Anonymous DH disabled, you can use:

x11vnc -create -vencrypt nodh:only -ssl

For Vino server, the only way to work around this server-side is to disable encryption with the following command on the command-line:

gsettings set org.gnome.Vino require-encryption false

I do not see how we can do better for Vino, unfortunately.

Client-side, I will probably throw up a message that AnonTLS is no longer supported if I find the app is running on Android 6 and newer.

**Comment by mzhou** [comment]
Awesome. Thanks for the response and investigation.
I'm happy to help test on Marshmallow if you ping this thread once the code is merged.
